### PR TITLE
Taught quick-lint-js about number literals like 1_234_567.

### PR DESCRIPTION
**Pull request by LeeWannacott**

### CODE REVIEW

File: src/lex.cpp [hunk 1]
```diff
@@ -662,7 +662,13 @@ void lexer::parse_number() {
         source_code_span(garbage_begin, garbage_end));
     input = garbage_end;
   };
-
+  input = this->parse_underscore(input);
+  bool has_underscore = *input == '_';
```
Comment (strager) [inline src/lex.cpp:666]: When is `has_underscore` true? The code seems to work without lines 666-671. What is the purpose of these lines? I thought `parse_underscore` has its own loop, so why are we calling it again?

**Review by strager — CHANGES_REQUESTED** [review]
Your code looks functional, but some of it looks redundant. Can you help me understand some of your decisions? Ideally, I'd like to see tests which demonstrate why some pieces of code (like lines 666-671) are necessary.

File: test/test-lex.cpp [hunk 2]
```diff
@@ -98,6 +98,7 @@ TEST(test_lex, lex_numbers) {
 
   check_single_token(u8"0n", token_type::number);
   check_single_token(u8"123456789n", token_type::number);
+  check_single_token(u8"1_234_567", token_type::number);
```
[the remark this answers is not in the document]
Comment (LeeWannacott) [reply test/test-lex.cpp:101]: Trailing underscore should not be allowed and should throw an error like: ' SyntaxError: Numeric separators are not allowed at the end of numeric literals'.  3e123_123_123 should be allowed.

File: src/lex.cpp [hunk 3]
```diff
@@ -662,7 +662,13 @@ void lexer::parse_number() {
...
+  if (has_underscore) {
+    input += 1;
+    input = this->parse_underscore(input);
+    has_underscore = true;
+  }
   input = this->parse_decimal_digits(input);
```
[the remark this answers is not in the document]
Comment (LeeWannacott) [reply src/lex.cpp:672]: True, I was a bit confused by the code at the time. I feel like I understand it better now.

File: src/lex.cpp [hunk 4]
```diff
@@ -651,7 +651,7 @@ void lexer::parse_binary_number() {
 }
 
 void lexer::parse_number() {
-  QLJS_ASSERT(this->is_digit(this->input_[0]) || this->input_[0] == '.');
+  QLJS_ASSERT(this->is_digit(this->input_[0]) || this->input_[0] == '.' || this->input_[0] == '_');
```
[the remark this answers is not in the document]
Comment (LeeWannacott) [reply src/lex.cpp:654]: A number can't start with a _ in JS. I got confused thinking that input[0] changes as input is incremented but it's not a loop. So I will remove this.

File: src/lex.cpp [hunk 5]
```diff
@@ -709,6 +715,18 @@ void lexer::parse_number() {
   this->input_ = input;
 }
 
+// TODO: Handle edge cases: 1__234_567 and 1_234_567_ should throw an error.
+const char8* lexer::parse_underscore(const char8* input) noexcept {
+  while (is_digit(*input)) {
+    input += 1;
+    bool has_underscore = *input == '_';
+    if (has_underscore) {
```
[the remark this answers is not in the document]
Comment (LeeWannacott) [reply src/lex.cpp:723]: Okay

File: src/lex.cpp (hunk 1 above)
Comment (LeeWannacott) [reply src/lex.cpp:666]: Yeah, it's unnecessary.

**Pull request closed without merging**